File: MiniDB_Projects/Team_BlueLeaf/src/txn/lock_manager.cpp
#include "txn/lock_manager.h"

#include <stack>

namespace minidb {

bool LockManager::compatible(const std::string& key, TxId tx, LockMode mode) {
    auto it = table_.find(key);
    if (it == table_.end()) return true;
    for (const Request& r : it->second) {
        if (!r.granted || r.tx == tx) continue;
        // S/S is fine; anything involving an X conflicts.
        if (mode == LockMode::EXCLUSIVE || r.mode == LockMode::EXCLUSIVE) return false;
    }
    return true;
}

bool LockManager::has_cycle(TxId start) {
    std::stack<TxId> st;
    std::unordered_set<TxId> seen;
    auto begin = waits_for_.find(start);
    if (begin != waits_for_.end())
        for (TxId n : begin->second) st.push(n);
    while (!st.empty()) {
        TxId u = st.top(); st.pop();
        if (u == start) return true;        // path returned to start -> cycle
        if (seen.count(u)) continue;
        seen.insert(u);
        auto it = waits_for_.find(u);
        if (it != waits_for_.end())
            for (TxId v : it->second) st.push(v);
    }
    return false;
}

void LockManager::acquire(TxId tx, const std::string& key, LockMode mode) {
    std::unique_lock<std::mutex> lk(mu_);
    auto& q = table_[key];

    // Already hold a lock on this key?
    for (Request& r : q) {
        if (r.tx == tx && r.granted) {
            if (r.mode == LockMode::EXCLUSIVE || mode == LockMode::SHARED) return;  // sufficient
            // Upgrade S -> X: wait until this tx is the only granted holder.
            for (;;) {
                std::unordered_set<TxId> blockers;
                for (const Request& o : q)
                    if (o.granted && o.tx != tx) blockers.insert(o.tx);
                if (blockers.empty()) { r.mode = LockMode::EXCLUSIVE; return; }
                waits_for_[tx] = blockers;
                if (has_cycle(tx)) { waits_for_.erase(tx); throw DeadlockException(tx); }
                cv_.wait(lk);
            }
        }
    }

    // New request.
    q.push_back(Request{tx, mode, false});
    for (;;) {
        if (compatible(key, tx, mode)) {
            for (Request& r : q) if (r.tx == tx && !r.granted) { r.mode = mode; r.granted = true; }
            held_[tx].insert(key);
            waits_for_.erase(tx);
            cv_.notify_all();
            return;
        }
        std::unordered_set<TxId> blockers;
        for (const Request& o : q)
            if (o.granted && o.tx != tx &&
                (mode == LockMode::EXCLUSIVE || o.mode == LockMode::EXCLUSIVE))
                blockers.insert(o.tx);
        waits_for_[tx] = blockers;
        if (has_cycle(tx)) {
            q.remove_if([tx](const Request& r) { return r.tx == tx && !r.granted; });
            waits_for_.erase(tx);
            cv_.notify_all();
            throw DeadlockException(tx);
        }
        cv_.wait(lk);
    }
}

void LockManager::release_all(TxId tx) {
    std::lock_guard<std::mutex> lk(mu_);
    for (auto& [key, q] : table_) q.remove_if([tx](const Request& r) { return r.tx == tx; });
    held_.erase(tx);
    waits_for_.erase(tx);
    for (auto& [t, s] : waits_for_) s.erase(tx);
    cv_.notify_all();
}

} // namespace minidb

File: MiniDB_Projects/Team_walError/src/minidb/lock_manager.py
"""lock_manager.py — a 2PL lock manager with wait-for-graph deadlock detection.

Locks are taken on opaque string "resources" (MiniDB uses table names, but the
manager is agnostic). Two modes:
    SHARED    (read)  — many transactions may hold it together
    EXCLUSIVE (write) — held by at most one transaction, conflicts with all

Compatibility:
              held SHARED   held EXCLUSIVE
    want S       yes             no
    want X        no             no

Blocking + deadlock: `acquire` blocks (on a condition variable) until the lock is
grantable. Before blocking it records the request in a wait-for graph and runs a
cycle check; if granting would create a deadlock, the *requesting* transaction is
chosen as the victim and a DeadlockError is raised instead of blocking forever.
Lock upgrade (a transaction holding S requesting X) is supported when it is the
sole holder.

Thread-safe: all state is guarded by a single Condition.
"""

from __future__ import annotations

import threading
from collections import defaultdict
from enum import Enum


class LockMode(Enum):
    SHARED = "S"
    EXCLUSIVE = "X"


class DeadlockError(Exception):
    """Raised in the victim transaction when acquiring a lock would deadlock."""

    def __init__(self, txn_id: int, resource: str) -> None:
        super().__init__(f"deadlock detected: txn {txn_id} aborted on {resource!r}")
        self.txn_id = txn_id
        self.resource = resource


def _conflict(a: LockMode, b: LockMode) -> bool:
    return a is LockMode.EXCLUSIVE or b is LockMode.EXCLUSIVE


class LockManager:
    def __init__(self) -> None:
        self._cond = threading.Condition()
        self._granted: dict[str, dict[int, LockMode]] = defaultdict(dict)
        self._waiting: dict[str, list[tuple[int, LockMode]]] = defaultdict(list)

    # --- public API --------------------------------------------------------

    def acquire(self, txn_id: int, resource: str, mode: LockMode) -> None:
        """Acquire `mode` lock on `resource` for `txn_id`, blocking if needed.

        Raises DeadlockError if waiting would create a cycle.
        """
        with self._cond:
            held = self._granted[resource].get(txn_id)
            if held is LockMode.EXCLUSIVE or held is mode:
                return  # already hold an equal-or-stronger lock
            while True:
                if self._can_grant(resource, txn_id, mode):
                    self._granted[resource][txn_id] = mode
                    self._remove_waiting(resource, txn_id)
                    self._cond.notify_all()
                    return
                # cannot grant yet -> record the wait and look for a deadlock
                if (txn_id, mode) not in self._waiting[resource]:
                    self._waiting[resource].append((txn_id, mode))
                if self._creates_cycle(txn_id):
                    self._remove_waiting(resource, txn_id)
                    self._cond.notify_all()
                    raise DeadlockError(txn_id, resource)
                self._cond.wait()

    def release(self, txn_id: int, resource: str) -> None:
        with self._cond:
            self._granted[resource].pop(txn_id, None)
            self._remove_waiting(resource, txn_id)
            self._cond.notify_all()

    def release_all(self, txn_id: int) -> None:
        """Release every lock held or requested by a transaction (commit/abort)."""
        with self._cond:
            for holders in self._granted.values():
                holders.pop(txn_id, None)
            for resource in list(self._waiting):
                self._waiting[resource] = [
                    (t, m) for (t, m) in self._waiting[resource] if t != txn_id
                ]
            self._cond.notify_all()

    # --- introspection (used by tests / demos) -----------------------------

    def holders(self, resource: str) -> dict[int, LockMode]:
        with self._cond:
            return dict(self._granted[resource])

    def wait_for_graph(self) -> dict[int, set[int]]:
        with self._cond:
            return self._build_wait_for()

    # --- internals (call only while holding self._cond) --------------------

    def _can_grant(self, resource: str, txn_id: int, mode: LockMode) -> bool:
        others = {t: m for t, m in self._granted[resource].items() if t != txn_id}
        if not others:
            return True
        if mode is LockMode.EXCLUSIVE:
            return False  # exclusive needs to be the sole holder
        # want SHARED: ok only if everyone else also holds SHARED
        return all(m is LockMode.SHARED for m in others.values())

    def _remove_waiting(self, resource: str, txn_id: int) -> None:
        self._waiting[resource] = [
            (t, m) for (t, m) in self._waiting[resource] if t != txn_id
        ]

    def _build_wait_for(self) -> dict[int, set[int]]:
        graph: dict[int, set[int]] = defaultdict(set)
        for resource, waiters in self._waiting.items():
            holders = self._granted[resource]
            for tw, mw in waiters:
                for h, hm in holders.items():
                    if h != tw and _conflict(mw, hm):
                        graph[tw].add(h)
        return graph

    def _creates_cycle(self, start: int) -> bool:
        graph = self._build_wait_for()
        # DFS looking for a path from `start` back to `start`
        seen: set[int] = set()
        stack = list(graph.get(start, ()))
        while stack:
            node = stack.pop()
            if node == start:
                return True
            if node in seen:
                continue
            seen.add(node)
            stack.extend(graph.get(node, ()))
        return False
